Lambda support

haozhun opened the pull request.

haozhun commented:
As I think about how `array_reduce` and `reduce` will work. I think it makes more sense to put lambdas at the end of function arguments. It would make sense to make the decision now so that everything is consistent.

Function | Lambda first | Lambda last
-------- | ------------ | -------------
apply           | `(T->U, T)                        -> U`        | `(T, T->U) -> U`
array_transform | `(T->U, array(T))                 -> array(U)` | `(array(T), T->U)                 -> array(U)`
array_filter    | `(T->boolean, array(T))           -> array(T)` | `(array(T), T->boolean)           -> array(T)`
array_reduce    | `((T,S)->S, S->R, array(T), S)    -> R`        | `(array(T), S, (T,S)->S, S->R)    -> R`
reduce          | `((T,S)->S, (S,S)->S, S->R, T, S) -> R`        | `(T, S, (T,S)->S, (S,S)->S, S->R) -> R`

martint commented:
That sounds reasonable. It's the same way I'm modeling operations in the new plan expression tree (e.g., a filter is `filter(source, lambda)`.)

However, I'd call the functions just `transform`, `filter` and `reduce` (what's the difference between reduce and array_reduce?)

electrum commented:
How about include/exclude instead of filter?

The name filter isn't self descriptive (it's ambiguous) and has the opposite behavior of the common English usage "filter out". Also, it's often useful to exclude things without inverting the predicate. I miss this with Java streams (especially with method references).

Ruby uses select/reject, which are good names, but not for SQL (obviously).

electrum commented:
I realize it's widely used in functional programming, but that's not the background of the average SQL user. I've always thought it's a bad name due to being ambiguous. C# uses "where" in LINQ.

It also doesn't help with having an inverse version. Is there another pair of names that works?

electrum commented:
I disagree "include" looks weird there. Only because you're used to the functional terms. Calling it "where" would be even better since it's 1:1 and avoids ambiguity.

martint commented:
But `where` can't be used as a function name since it's a reserved keyword (it's the same function for arrays and for multisets of rows, which is what a table produces)

cberner commented:
Why do we need to allocate symbols in the `TranslationMap`? I haven't thoroughly reviewed the last commit yet, but it seems like it's rewriting the argument names and their references into unique symbols. Is that necessary? It seems like that compiler and interpreter could just use the original names

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java:
```diff
@@ -50,6 +51,9 @@
     private final Map<Expression, Expression> expressionToExpressions = new HashMap<>();
 
+    // memoize translated expressions so that the same output expression will be produced for each input expression
+    private final IdentityHashMap<Expression, Expression> translatedExpressions = new IdentityHashMap<>();
```
I think that to make this work you would need to make `TranslationMap` to be immutable. Then you could have `CachingTranslationMap` around it. Alternatively, you could clean that map every time somebody calls any of `put*` method, but this kind of approach is more error-prone to me.

cberner replied:
Is this memoization required for correctness, or is it a performance optimization?

sopel39 replied:
does it have to be `IdentityHashMap`?

haozhun replied:
Yes. But I'm removing it so it doesn't matter any more.

cberner commented:
These look good:
* Add context to ExpressionAnalyzer.Visitor
* Add context to SymbolToInputRewriter
* Add context to TryExpressionExtractor
* Make SymbolAllocator available to TranslationMap

cberner commented:
@martint should review `Make TranslationMap memoize translated expressions`

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,98 @@
+        return typeSignatureSupplier.apply(ImmutableList.of());
+    }
+
+    public TypeSignature getTypeSignature(List<Type> boundTypeParameters)
```
what is the intention of `boundTypeParameters`? Is this an ordered list? Is this intended to be used as a lambda type provider?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,98 @@
+                .collect(ImmutableCollectors.toImmutableList());
+    }
+
+    public static Optional<List<Type>> toTypes(List<TypeSignatureProvider> typeSignatureProviders, TypeManager typeManager)
```
does it have to operate on lists? Is seems more natural for `toTypes` function to operate on singular elements. You also wouldn't need `hasDependency` field

haozhun replied:
Thank you for the suggestion. Done.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,98 @@
+
+public class TypeSignatureProvider
+{
+    private final boolean hasDependency;
```
shouldn't `hasDependency` be an integer?

haozhun replied:
The only way an integer would be useful here is so that validation can be done before calling the resolver. However, this would require the number be made available at construction time. I don't find the validation useful right now as I explained above. We can certainly change this if a use case comes up later that makes a number useful.

sopel39 replied:
From this commit it's hard to get context how `TypeSignatureProvider` is intended to be used. I see that `hasDependency` flag determines how this object is being treated by callers. Correct me if I am wrong, but do I understand it correctly that `hasDependency` determines if `TypeSignatureProvider` represents lambda or plain type? If that is the case, maybe it would be better to make `TypeSignatureProvider` a clean interface and add two implementations: e.g: `BasicTypeSignatureProvider` and `LambdaSignatureProvider`?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,98 @@
+        this.typeSignatureSupplier = ignored -> typeSignature;
+    }
+
+    public TypeSignatureProvider(Function<List<Type>, TypeSignature> typeSignatureSupplier)
```
It seems that `typeSignatureSupplier` parameter is named as such because `TypeSignatureProvider` name is already taken. Is there a better (more meaningful, e.g: `typeSignatureResolver`) name for `typeSignatureSupplier` parameter?

haozhun replied:
Done. Thank you.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:
```diff
@@ -198,6 +198,7 @@
 import static com.facebook.presto.operator.aggregation.MinByNAggregationFunction.MIN_BY_N_AGGREGATION;
```
is it possible to split this commit into smaller ones. It's quite large chunk of code.

kokosing replied:
You could separate changes into separate commits have for:
 - `Apply` funciton
 - code generation 
 - semantic analyzer

haozhun replied:
I split out a few refactoring and preparatory commits.

I didn't split things out following suggestions from @kokosing. To me, it's more important to make sure that two unrelated changes that mix together doesn't go in the same commit.

`Apply` function is largely unrelated to the rest of the code. It's act as a test mostly. As a result, I don't see how splitting that out can make code review easier. Additionally, without the apply function, code generation and semantic analyzer will basically go untested.

I didn't spend the time to split out code generation and semantic analyzer for the same reason. 1. Additional tests will be needed if I do so. 2. They are not really mixed together anyways.

kokosing replied:
Thanks!

cberner commented:
This one looks good to me: `Introduce TypeSignatureProvider to SignatureBinder`. See @sopel39's comments on naming though. I don't really have an opinion on that.

Back to you for refactorings that we discussed offline

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,98 @@
+        this.typeSignatureSupplier = requireNonNull(typeSignatureSupplier, "typeSignatureSupplier is null");
+    }
+
+    public boolean hasDependency()
```
Can you help me understand why you think this would be better called `isStructural`? Can you also share your concern about why you don't like the original name?

Generally, structural types means array, map, row, etc. I don't really think of function type as a structural type. Regardless, given structural type also includes array, map, row, `isStructural` doesn't seem like a good name?

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,98 @@
+        return hasDependency;
+    }
+
+    public TypeSignature getTypeSignature()
```
I don't understand your argument here. As for the return type of this method, I believe most callers are not interested in types with dependencies. As a result, returning an Optional does not seem useful. In addition, this will make callers more complicated. I prefer leaving this as is.

sopel39 replied:
You could have two methods:
```
Optional<TypeSignature> getTypeSignature(...)
TypeSignature getTypeSignature(...)
```
The second would enforce result to be present. Maybe I don't have the full context, but I find it strange to treat no-arguments (`hasDependency=false`) case any special way (e.g: have two `getTypeSignature` methods, one with list the other without). Why no-arguments case is enforced while n-arguments case is not? What stops you now from calling `getTypeSignature` directly with empty list?

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,98 @@
+
+    public TypeSignature getTypeSignature()
+    {
+        checkState(!hasDependency);
```
Doing the validation in the other function as you suggested would require having an integer in the type signature provider. I don't have such an integer. As a result, I don't do the check.

Given the check can be done in the `resolver`, I don't find it particularly important to have the number available so that the check can be done.

sopel39 replied:
You don't have such integer because you don't know how many arguments will be applied? Does it mean that `Function<List<Type>, TypeSignature> typeSignatureResolver)` is a vargarg resolver?

In any case it seems strange to have `TypeSignature getTypeSignature()` method just for plain no-arg types. You could have no-arg lambdas also but then you can't use use them in places where `TypeSignature getTypeSignature()` is called. Maybe plain types deserve to be treated separately (e.g: have separate provider)?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,84 @@
+        this.typeSignatureResolver = requireNonNull(typeSignatureResolver, "typeSignatureResolver is null");
+    }
+
+    public boolean hasDependency()
```
As to the name of `hasDependency` flag. Is it dependency on parameters or some other stuff (e.g: where lamba is applied). If it is possible to make the name more meaningful (e.g: `requiresArgumentTypes`?) that would be great.

I agree `isStructural` wouldn't be any better.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -296,6 +314,14 @@ protected Type visitCurrentTime(CurrentTime node, StackableAstVisitorContext<Con
         @Override
         protected Type visitSymbolReference(SymbolReference node, StackableAstVisitorContext<Context> context)
         {
+            if (context.getContext().isInLambda()) {
```
`StackableAstVisitorContext` contains information (path) about AST nodes from the current node up to the root. I think you could leverege that instead of writing and your own context.

I could image that as:
```
if (context.isIn(LambdaExpression.class)) {
````

kokosing replied:
`StackableAstVisitorContext` was introduced in order to provide an information about placement of currently processed node in AST. I think it has a value to extend to support such generic problems to know if this node is wrapped in another node (even if there are nodes between). Thanks to that it could be then used in other such cases as well.

Nevertheless, I understand that you will need your own context anyway.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -749,12 +817,35 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
+                }
+                else {
+                    Type actualType = typeManager.getType(argumentTypes.get(i).getTypeSignature());
+                    coerceType(expression, actualType, expectedType, format("Function %s argument %d", function, i));
```
This method `visitFunctionCall` has already about 200 lines of code. Can you please extract some methods with meaningful names. It would make this code much more readable.

haozhun replied:
I simplified this method to 130 lines. It's 90 lines if you don't count the window function validation section. I'll leave it as is for now.

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,84 @@
+    {
+        return typeSignatures.stream()
+                .map(TypeSignatureProvider::new)
+                .collect(ImmutableCollectors.toImmutableList());
```
Done. Thank you.

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,84 @@
+        return typeSignatureResolver.apply(boundTypeParameters);
+    }
+
+    public static List<TypeSignatureProvider> fromTypes(List<? extends Type> types)
```
I don't think we have this convention. We do generally put methods that are essentially constructors near constructors. But these two methods aren't really like that given they produce a list.

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    public TypeSignature getTypeSignature(List<Type> boundTypeParameters)
+    {
```
Done

haozhun commented:
By the way, it took me longer than expected to resolve all the review commits from @kokosing and @sopel39. I left this pull request assigned to myself while I'm incrementally addressing comments. I'm open to you leaving comments while the pull request is in this state as long as you understand what you are observing is intermediate.

haozhun commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:
```diff
@@ -619,25 +621,25 @@ public Signature resolveFunction(QualifiedName name, List<TypeSignature> paramet
     }
 
-    private Optional<Signature> matchFunctionExact(List<SqlFunction> candidates, List<Type> actualParameters)
+    private Optional<Signature> matchFunctionExact(List<SqlFunction> candidates, List<TypeSignatureProvider> actualParametersTypeSignatureProvider)
```
Good point

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/TypeSignatureProvider.java:
```diff
@@ -0,0 +1,84 @@
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class TypeSignatureProvider
```
I think your suggestions here are reasonable. However, I don't really see a common interface between the two. Essentially, there will be two functions in the interface, but only one of them is implemented each. I think this class is simple enough right now. We can totally change this if and when this becomes more complicated.

haozhun commented:
I removed memoization from `TranslationMap`.

The last 5 commits are new. Back to you, @cberner.

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/gen/TryCodeGenerator.java:
```diff
@@ -107,6 +110,7 @@ public static MethodDefinition defineTryMethod(
 
         method.comment("Try projection: %s", innerRowExpression.toString());
         method.getBody()
+                .append(wasNull.set(constantBoolean(false)))
```
Is this a fix for another issue? It seems unrelated to the rest of the refactoring

haozhun replied:
No. This is not a fix for another issue.

This is just part of changing
```
RetType f(..., boolean wasNull) {
  ...
}
```
to
```
RetType f(...) {
  boolean wasNull = false;
  ...
}
```

In the former case, invocation always pass `false` for `wasNull`.

Writing it without ` = false` does not work because byte code generation generally assumes that `wasNull` is initially false.
```
RetType f(...) {
  boolean wasNull;
  ...
}
```

cberner commented:
I'm still reviewing `Add lambda expression support`, but the first 8 commits look good.

cberner commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestApplyFunction.java:
```diff
@@ -0,0 +1,99 @@
+    public void testNestedLambda()
+            throws Exception
+    {
+        assertFunction("apply(x -> apply(y -> apply(z -> z * 5, y * 3) + 1, x + 7) * 2, 11)", INTEGER, 542);
```
Add a test like `apply(x -> apply(y -> x * x, x), 2)`. Does the capturing work?

cberner commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestApplyFunction.java:
```diff
@@ -0,0 +1,99 @@
+import static com.facebook.presto.spi.type.VarcharType.createUnboundedVarcharType;
+import static com.facebook.presto.spi.type.VarcharType.createVarcharType;
+
+public class TestApplyFunction
```
Add test for NULL handling, like `apply(x -> x + 1, NULL)`. Also, test functions that rely on `ConnectorSession` or that are instance methods

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java:
```diff
@@ -41,6 +45,7 @@
     // all expressions are rewritten in terms of fields declared by this relation plan
     private final RelationPlan rewriteBase;
     private final Analysis analysis;
+    private final IdentityHashMap<LambdaArgumentDeclaration, Symbol> lambdaDeclarationToSymbolMap;
```
Why is this separate from `expressionToSymbols`? Also, why isn't it copied in `copyMappingsFrom` and `putExpressionMappingsFrom`?

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -20,6 +20,7 @@
 import com.facebook.presto.spi.type.TypeSignature;
 import com.facebook.presto.spi.type.TypeSignatureParameter;
 import com.facebook.presto.sql.analyzer.TypeSignatureProvider;
+import com.facebook.presto.type.FunctionType;
```
Add tests to `TestSignatureBinder`. I would test things like:
`f(function<int, int>)` binds to `f(x -> cast(x as smallint))`
`f(function<int, int>)` does not binds to `f(x -> cast(x as bigint))`
`f(function<int, int>...)` to make sure varargs works
`f(function<T, U>, function<U, V>, T):V` binds to `f(x -> x + 1.0, y -> cast(y as varchar), 0)`
Also check things with parametric types, like `f(function<T, array<T>>, T)`

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ApplyFunction.java:
```diff
@@ -0,0 +1,94 @@
+    public static Object apply(MethodHandle function, Object input)
+    {
+        try {
+            return function.invoke(input);
```
It works. I added test to cover this case. Right now, lambdas are always generated as nullable.

haozhun commented:
Addressed comments. Back to you, @cberner.

haozhun commented:
Travis passed.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/VarArgsToMapAdapterGenerator.java:
```diff
@@ -0,0 +1,94 @@
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner;
```
This should probably go in the "gen" package

haozhun replied:
Done

martint commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/LambdaExpression.java:
```diff
@@ -22,27 +22,29 @@
-    private final List<String> arguments;
+    private final List<LambdaArgumentDeclaration> arguments;
     private final Expression body;
+    private final boolean bodyContainsSymbolReferences;
```
This seems weird. Symbols are a planner concept, so, ideally, the AST shouldn't have to be aware of them.

Also, it's something that can only be determined at analysis/planning time, not during parsing.

haozhun replied:
I removed it because it seems no longer necessary.

martint commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/LambdaArgumentDeclaration.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.Objects;
+import java.util.Optional;
+
+public class LambdaArgumentDeclaration
```
Why do we need a dedicated node type for lambda arguments?

haozhun replied:
Yes, we do. We talked about before. This is needed to make associating symbols and types with lambdas easy. The alternative is to use QualifiedNameReference and SymbolReference here. Using a dedicated types seems preferable here.

cberner commented on presto-main/src/test/java/com/facebook/presto/metadata/TestSignatureBinder.java:
```diff
@@ -986,11 +988,82 @@ public void testUnknownCoercion()
+                .produces(BoundVariables.builder()
+                        .setTypeVariable("T", INTEGER)
+                        .build());
+    }
```
How about varargs? I'd test a function `apply_all(T, function<T, T>...) -> T` and make sure it can bind to `apply_all(1, x -> x + 1)` and `apply_all(1, x -> x + 1, y -> y + 1)`

haozhun replied:
I don't see a use case. I added the test anyways.

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -256,20 +262,36 @@ private boolean appendConstraintSolvers(
+        if (FunctionType.NAME.equals(formalTypeSignature.getBase())) {
+            List<TypeSignature> formalTypeParameterTypeSignatures = formalTypeSignature.getTypeParametersAsTypeSignatures();
+            resultBuilder.add(new FunctionSolver(
+                    getLambdaArgumentTypeSignatures(formalTypeSignature),
```
I don't think "lambda" is the right name to use in `SignatureBinder`. Not all `FunctionType`s are lambdas, since I could register a `gen_function() -> function<integer, integer>` that generates functions

cberner replied:
In fact, we should add a test for a function that generates functions to make sure you can do something like `apply(1, gen_test_function())`.

haozhun replied:
Right now, lambda is the only supported instance of function type. While supporting functions that return non-generic function types is likely simple, I don't see a use case.

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -742,14 +776,44 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<V
+                    Expression lambdaBody = lambdaExpression.getBody();
+                    List<LambdaArgumentDeclaration> lambdaArguments = lambdaExpression.getArguments();
+
+                    ExpressionAnalyzer innerExpressionAnalyzer = new ExpressionAnalyzer(
```
Done

haozhun commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/LambdaExpression.java:
```diff
@@ -52,6 +54,11 @@ public Expression getBody()
         return body;
     }
 
+    public boolean isBodyContainsSymbolReferences()
```
Removed

haozhun commented on presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:
```diff
@@ -82,7 +84,7 @@ public SignatureBinder(TypeManager typeManager, Signature declaredSignature, boo
     }
 
-    public Optional<Signature> bind(List<? extends Type> actualArgumentTypes)
+    public Optional<Signature> bind(List<? extends TypeSignatureProvider> actualArgumentTypes)
```
Done.

haozhun commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:
```diff
@@ -527,6 +527,7 @@ public WindowFunctionSupplier load(SpecializedFunctionKey key)
                 .function(DECIMAL_MOD_FUNCTION)
                 .functions(DECIMAL_ROUND_FUNCTIONS)
                 .function(DECIMAL_TRUNCATE_FUNCTION)
+                .function(APPLY_FUNCTION)
```
Done.

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/gen/PreGeneratedExpressions.java:
```diff
@@ -0,0 +1,37 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class PreGeneratedExpressions
```
I think differently on this topic.

The implementation of lambda expression is a piece of code that invokes the MethodHandle stored as a static final field corresponding to a method pre-generated for the lambda expression. Implementation of try expression is similar. That's why I don't like "ExpressionImplementations", "CompiledExpressions" or something along those lines.

Since the introduction of try, the generation of byte code for expression is two steps: a prepare phase and a main phase. 

This class is introduced to avoid passing multiple maps around between the prepare phase and the main phase. It helps encapsulate those maps into an entity. It serves a very specific function. And the current name explains that specific function.

If we ever need to generalize the function of this class, we can rename it at that point. We will be able to come up with a name appropriate to the function this class serves at that point. A few things around this class will likely be re-organized as well. Names that are not appropriate right now might be appropriate under the new context.

For the above reasons combined, I prefer using `PreGeneratedExpressions` as the name of this class.

wenleix replied:
@haozhun We should add comment for this :)

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/gen/PreGeneratedExpressions.java:
```diff
@@ -0,0 +1,37 @@
+
+public class PreGeneratedExpressions
+{
+    private final Map<CallExpression, MethodDefinition> tryMethodMap;
```
See comment above.

haozhun commented:
@cberner To answer your concerns:

* Right now, lambda is the only supported instance of function type. While supporting functions that return non-generic function types is likely simple, I don't see a use case.
* In terms of tests, lambda is only exposed through functions. And I believe the hidden APPLY function in this pull request is thoroughly tested. I don't really see other potential areas that are likely to cause correctness bugs. Tests of same caliber will accompany other lambda functions I'm about to introduce.
